src/haxelib/VersionData.hx

```diff
@@ -0,0 +1,89 @@
+package haxelib;
+
+/** Abstract enum representing the types of Vcs systems that are supported. **/
+@:enum abstract VcsID(String) to String {
+	var Hg = "hg";
+	var Git = "git";
+
+	/** Returns `true` if `s` constitutes a valid VcsID **/
+	public static function isValid(s:String) {
+		return s == Hg || s == Git;
+	}
+
+	/** Returns `s` as a VcsID if it is valid, otherwise throws an error. **/
+	public static function ofString(s:String):VcsID {
+		if (s == Git)
+			return Git;
+		else if (s == Hg)
+			return Hg;
+		else
+			throw 'Invalid VscID $s';
+	}
+}
+
+/** Class containing repoducible git or hg library data. **/
+@:structInit @:publicFields
+class VcsData {
+	/** url from which to install **/
+	var url:String;
+	/** Commit hash **/
+	@:optional
+	var ref:Null<String>;
+	/** The git tag or mercurial revision **/
+	@:optional
+	var tag:Null<String>;
+	/** Branch **/
+	@:optional
+	var branch:Null<String>;
+	/**
+		Sub directory in which the root of the project is found.
+
+		Relative to project root
+	**/
+	@:optional
+	var subDir:Null<String>;
+}
+
+/** Data required to reproduce a library version **/
+enum VersionData {
+	Haxelib(version:SemVer);
+	VcsInstall(version:VcsID, vcsData:VcsData);
+}
+
+class VersionDataHelper {
+	public static function extractVersion(versionInfo:String):VersionData {
+		try {
+			return Haxelib(SemVer.ofString(versionInfo));
+		} catch (_:String) {}
+
+		try {
+			var data = getVcsData(versionInfo);
+			return VcsInstall(data.type, data.data);
+		} catch (e:String) {
```

Comment:
I don't like how we catch a string exception and then ignore it by throwing a different string exception. I think functions like `getVcsData` should just return an `Option` in cases like that. But yes I know you didn't write this, so we can change this later.

Reply:
Ah I probably did write this, just at an earlier point...

Reply:
I'll change it because this is the only place where the function is called now, so shouldn't be too hard to refactor.